### PR TITLE
Add support for ASP.NET Core 3

### PR DESCRIPTION
**Pull request by alsami**
This is adding support for ASP.NET Core 3, introducing breaking changes for older versions of ASP.NET Core but still supports them.

* add multitargeting to framework `netstandard2.0` and `netcoreapp3.0`
* update `Autofac*` peer dependencies, some of them currently previews from autofac-myget
* change multitenant-container accessor to require an instance of `IContainer` that can be passed to it and is now only being used in `AutofacMultitenantServiceProviderFactory`
* create `AutofacMultitenantServiceProviderFactory` and capture the create service-provider in a closure registration
* move test-host configuration into `TestServerFixture`
* adapt changes in tests

Related to #9 and [#36](https://github.com/autofac/Autofac.Extensions.DependencyInjection/issues/36).

### CODE REVIEW

**Comment by alsami** [comment]
Hey @tillig, @nblumhardt, @alexmg 

I found a hacky work-around that semi-works. Some tests are failing when doing concurrent-request. This might not be related to the changes here but rather to the changes made to ASP.NET-Core.

I have asked [David Fowler on Twitter](https://twitter.com/alsami_x5A/status/1164552006746419202) how we could receive the actual root-container being created by the [HostBuilder](https://github.com/aspnet/Extensions/blob/f8d29bc1344364c7444bec17049df653f82760c7/src/Hosting/Hosting/src/HostBuilder.cs#L233). Somehow it's not possible and the best we can get is the root-lifetime-scope used be the instance of `IContainer`.

image

Problem is, the `MultitenantContainer` doesn't allow an instance of `LifetimeScope` to be passed in, even when the tag is `root`.

I definitely need some support here. Can anybody help out there?

**Comment by saber-wang** [comment]
So I use a custom ServiceProviderFactory in my project.
```
public class AutofacFactory : IServiceProviderFactory<ContainerBuilder>
    {
        MultitenantContainer ApplicationContainer = null;
        public ContainerBuilder CreateBuilder(IServiceCollection services)
        {
            services.AddAutofacMultitenantRequestServices(() => ApplicationContainer);
            services.AddSingleton<ITenantIdentificationStrategy, RequestParameterStrategy>();
            var builder = new ContainerBuilder();
            builder.Populate(services);
            return builder;
        }
        public IServiceProvider CreateServiceProvider(ContainerBuilder containerBuilder)
        {
            var container = containerBuilder.Build();
            ITenantIdentificationStrategy strategy;
            using (var scope = container.BeginLifetimeScope())
            {
                strategy = scope.Resolve<ITenantIdentificationStrategy>();
            }
            ApplicationContainer = new MultitenantContainer(strategy, container);
            //TODO ApplicationContainer.ConfigureTenant
            return new AutofacServiceProvider(ApplicationContainer);
        }
    }
```
Are there any other potential problems?

**Comment by alsami** [comment]
@saber-wang I'd kindly ask you to move your question to the issue.

@tillig @alexmg @nblumhardt 

I fixed the test. The problem was the registration of `ITenantAccessor` as a singleton in the test-services. The tests were probably not executed simultaneously before which is why this didn't happen. Now it's quite fast and concurrent tests were overriding each others expectations.

Ready to review and discuss the changes. Any suggestion on how to do it in a less hacky-way are welcome 😄

**Comment by saber-wang** [comment]
@alsami No problem-_-. I think I should look at the source code.

File: samples/Sandbox/Dependency.cs [hunk 1]
```diff
@@ -10,14 +9,14 @@ public class Dependency : IDependency, IDisposable
 
         public Dependency(ILogger<Dependency> logger)
         {
-            this._logger = logger;
+            _logger = logger;
```
Comment (tillig) [inline samples/Sandbox/Dependency.cs]: I don't mind little optimizations and things, but I _am_ a fan of putting `this.` in front of instance items for clarity. It also makes code reviews kinda large with unrelated changes. Something to consider.

Comment (tillig) [reply samples/Sandbox/Dependency.cs]: Maybe at some point I need to add updated .editorconfig and analyzer ruleset settings to fix this stuff. Otherwise there's going to be the same sort of code churn of add/remove/add/remove as different preferences switch these things. Hmmm.

File: samples/Sandbox/QueryStringTenantIdentificationStrategy.cs [hunk 2]
```diff
@@ -1,27 +1,24 @@
-﻿using System;
-using System.Linq;
-using Autofac.Multitenant;
+﻿using Autofac.Multitenant;
```
Comment (tillig) [inline samples/Sandbox/QueryStringTenantIdentificationStrategy.cs]: Good example of churn and larger reviews - this file didn't actually need to be touched at all to get things done.

Can you revert changes on files that didn't need to change so there's less to review here? Thanks.

**Review by tillig — CHANGES_REQUESTED** [review]
It currently says 29 files changed but in a lot of cases the files didn't need to change. Can we revert the changes on files that didn't need to change so there's less to review? Thanks.

File: samples/Sandbox/Dependency.cs (hunk 1 above)
Comment (alsami) [reply samples/Sandbox/Dependency.cs]: I am also a friend of this to be honest. I usually don't use underscore for the fields then though. I will readd them!

File: samples/Sandbox/QueryStringTenantIdentificationStrategy.cs (hunk 2 above)
Comment (alsami) [reply samples/Sandbox/QueryStringTenantIdentificationStrategy.cs]: I cleaned up unused imports :)

**Comment by alsami** [comment]
> It currently says 29 files changed but in a lot of cases the files didn't need to change. Can we revert the changes on files that didn't need to change so there's less to review? Thanks.

@tillig I have undone most of the changes and also the renaming of the projects. If you don't mind, I'd align the project names (src and test) to fit to the actual package name.

Please review again.

File: src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -24,12 +24,12 @@
...
 
 using System;
-using Autofac.Integration.AspNetCore.Multitenant;
 using Autofac.Multitenant;
+using Microsoft.AspNetCore.Hosting;
 using Microsoft.AspNetCore.Http;
 using Microsoft.Extensions.DependencyInjection;
 
-namespace Microsoft.AspNetCore.Hosting
+namespace Autofac.Integration.AspNetCore.Multitenant
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceCollectionExtensions.cs]: In the past we kept this in the Microsoft.AspNetCore.Hosting namespace intentionally so folks can discover it and don't have to include something extra. That's pretty common for libraries contributing to hosting, just like putting extensions in Microsoft.Extensions.Configuration is pretty common for libraries extending configuration.

File: src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantWebHostBuilderExtensions.cs [hunk 4]
```diff
@@ -25,8 +25,9 @@
 
 using System;
 using Autofac.Multitenant;
+using Microsoft.AspNetCore.Hosting;
 
-namespace Microsoft.AspNetCore.Hosting
+namespace Autofac.Integration.AspNetCore.Multitenant
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantWebHostBuilderExtensions.cs]: This should also stay in Microsoft.AspNetCore.Hosting.

File: src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj [hunk 5]
```diff
@@ -2,8 +2,8 @@
 
   <PropertyGroup>
     <Description>ASP.NET Core support for multitenant DI via Autofac.Multitenant.</Description>
-    <VersionPrefix>1.1.0</VersionPrefix>
-    <TargetFrameworks>netstandard1.3;netstandard2.0</TargetFrameworks>
+    <VersionPrefix>2.0.0</VersionPrefix>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj]: We can drop ASP.NET Core 1.x support, but we do still need to support ASP.NET Core 2.x. I'll mention more on that in a minute in the middleware.

File: src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs [hunk 6]
```diff
@@ -21,37 +19,20 @@ public static IServiceProvidersFeature CreateFeature(HttpContext context, IServi
...
-                                factoryParameter),
-                            typeof(IServiceProvidersFeature)),
-                        contextParameter,
-                        factoryParameter)
-                        .Compile();
-                default:
-                    throw new NotSupportedException(Properties.Resources.NoSupportedRequestServicesConstructorFound);
-            }
+
+            return Expression.Lambda<Func<HttpContext, IServiceScopeFactory, IServiceProvidersFeature>>(
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs]: The reason this was an expression that generates a lambda and not just a reference right in the middleware to get the request services feature is that the constructor signature for `RequestServicesFeature` changed between ASP.NET Core 2.0 and ASP.NET Core 2.1.

- ASP.NET Core 2.0: [`public RequestServicesFeature(IServiceScopeFactory scopeFactory)`](https://github.com/aspnet/Hosting/blob/e127224812d6fa30469f5c30f3f6bea43fb27b26/src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesFeature.cs#L18)
- ASP.NET Core 2.1: [`public RequestServicesFeature(HttpContext context, IServiceScopeFactory scopeFactory)`](https://github.com/aspnet/Hosting/blob/55e57af8156eab2c28c62b5c6c5473405cb1c411/src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesFeature.cs#L20)
- ASP.NET Core 3.0: [`public RequestServicesFeature(HttpContext context, IServiceScopeFactory scopeFactory)`](https://github.com/aspnet/AspNetCore/blob/4a42afc5aea63750638e118560d43db04bd9ccc2/src/Http/Http/src/Features/RequestServicesFeature.cs#L18)

I'd probably be OK if we switched to ASP.NET Core 2.1.x minimum here and dropped 2.0, but if we do that, this whole lambda thing isn't required anymore and we should just new-up the feature in the middleware rather than using this indirection thing.

If we keep ASP.NET Core 2.0 support, we'll still need the switch statement here based on parameter count or things for ASP.NET Core 2.0 folks will break.

File: test/Autofac.Integration.AspNetCore.Multitenant.Test/Autofac.Integration.AspNetCore.Multitenant.Test.csproj [hunk 7]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.2</TargetFramework>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
```
Comment (tillig) [inline test/Autofac.Integration.AspNetCore.Multitenant.Test/Autofac.Integration.AspNetCore.Multitenant.Test.csproj]: Just like with the main package, we may need to multitarget here.

File: src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj [hunk 8]
```diff
@@ -41,9 +41,7 @@
     <PackageReference Include="Microsoft.Net.RoslynDiagnostics" Version="2.6.1">
       <PrivateAssets>All</PrivateAssets>
     </PackageReference>
-    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta-62925-02" Condition="Exists('$(MSBuildThisFileDirectory)../../.git')">
-      <PrivateAssets>All</PrivateAssets>
-    </PackageReference>
+    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-19367-01" />
     <PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta008">
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj]: If you update the analyzers in the test project, you should keep the main project on the same version for consistency.

File: samples/Sandbox/Startup.cs [hunk 9]
```diff
@@ -1,65 +1,65 @@
...
-        public IConfiguration Configuration { get; }
+            app.UseEndpoints(builder => builder.MapControllers());
+        }
 
-        public void Configure(IApplicationBuilder app)
+        public void ConfigureServices(IServiceCollection services)
         {
-            app.UseMvc();
+            services
+                .AddAutofacMultitenantRequestServices(ConfigureMultitenantContainer)
```
Comment (tillig) [inline samples/Sandbox/Startup.cs]: I like this concept of having a callback. I think it's a good solution to working around how to get the built container into the factory and configuring the container. Slick.

File: src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace Autofac.Integration.AspNetCore.Multitenant
+{
+    /// <summary>
+    /// A factory for creating a <see cref="ContainerBuilder"/> and an <see cref="IServiceProvider" /> for usage with a <see cref="MultitenantContainer" /> in ASP.NET Core.
+    /// </summary>
+    public class AutofacMultitenantServiceProviderFactory : IServiceProviderFactory<ContainerBuilder>
+    {
+        private readonly Action<ContainerBuilder> _configurationAction;
+        private static IContainer container;
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs]: Statics are problematic. If we have to register something specific to support getting the container in and out of the services list, it'd be better to have some sort of GUID generated and use a keyed service.

File: src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        /// <summary>
+        /// Creates a container builder from an <see cref="IServiceCollection" />.
+        /// </summary>
+        /// <param name="services">The collection of services.</param>
+        /// <returns>A container builder that can be used to create an <see cref="IServiceProvider" />.</returns>
+        public ContainerBuilder CreateBuilder(IServiceCollection services)
+        {
+            services.AddSingleton(_ => AutofacMultitenantServiceProviderFactory.container);
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs]: If there's a way to avoid tying the built container to this factory (e.g., via closure) that'd be good. For example, you could add a property to the `ContainerBuilder.Properties` dictionary that includes the key of a registered container that could be resolved later. That way it's just in the builder and no ties at all to the factory.

File: src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs [hunk 12]
```diff
@@ -16,20 +15,24 @@ internal class MultitenantRequestServicesMiddleware
...
         private readonly RequestDelegate _next;
 
+        private readonly IContainer _container;
+
         /// <summary>
         /// Initializes a new instance of the <see cref="MultitenantRequestServicesMiddleware"/> class.
         /// </summary>
         /// <param name="next">The next step in the request pipeline.</param>
         /// <param name="multitenantContainerAccessor">A function that will access the multitenant container from which request lifetimes should be generated.</param>
+        /// <param name="container">TODO.</param>
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs]: Docs to fill out here.

File: src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs [hunk 13]
```diff
@@ -52,18 +55,21 @@ public async Task Invoke(HttpContext context)
                 this._contextAccessor.HttpContext = context;
             }
 
-            var container = this._multitenantContainerAccessor();
-            if (container == null)
+            var multitenantContainer =
+                this._multitenantContainerAccessor(_container);
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs]: Unfortunately, this won't work. You can't re-build the multitenant container on every request. It needs to be built exactly once at app startup time, before this middleware is invoked.

It can only be built once because if folks have singletons (e.g., instance per tenant) then those get disposed on every request here and that can't happen. The multitenant container needs to live for the app lifetime just like the root container for instance sharing to work.

It needs to happen before multitenant middleware (e.g., at app startup) because folks kick off background processes and things that share the container and they assume the service provider is ready to go. If it only gets built in request middleware, that assumption will break.

And, of course, it's really slow (relatively speaking) to build up a container, so putting that into a request pipeline is a pretty big perf hit.

File: src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        /// <returns>An <see cref="IServiceProvider" />.</returns>
+        public IServiceProvider CreateServiceProvider(ContainerBuilder containerBuilder)
+        {
+            if (containerBuilder == null) throw new ArgumentNullException(nameof(containerBuilder));
+
+            var rootContainer = containerBuilder.Build();
+
+            if (AutofacMultitenantServiceProviderFactory.container == null) AutofacMultitenantServiceProviderFactory.container = rootContainer;
+
+            return new AutofacServiceProvider(rootContainer);
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs]: The service provider that comes out of here needs to be based on the multitenant container so resolutions immediately after the provider is created will be based on multitenant needs. Multitenancy isn't just needed in the request pipeline.

**Review by tillig — CHANGES_REQUESTED** [review]
I think the overall pattern of providing a callback to the multitenant factory to configure the multitenant container is good, but the execution of the callback needs to happen right in the factory and put into the returned AutofacServiceProvider rather than put into the middleware. This may mean there needs to be an AutofacMultitenantServiceProvider that derives from the main AutofacServiceProvider but extends it to add a reference to the root app container AND the multitenant container.

Since the "lifetime scope" passed to the provider will be the multitenant container, the additional property (and extension method) will probably need to be to get the root container.

If there's a better way, I'm open to suggestions - I'm not saying "a new service provider class must happen!", I'm saying that's one way to fix it.

**Comment by tillig** [comment]
> I have undone most of the changes and also the renaming of the projects. If you don't mind, I'd align the project names (src and test) to fit to the actual package name.

Let's leave that for now. This integration library was built following the patterns we'd established in the Autofac.Mvc, Autofac.WebApi, Autofac.Owin, and other integration packages - where the package name is short, but the assembly inside has 'Integration' in the name. I'm not 100% glued to that, but there's not a great reason to change it, either, and that's a pretty breaking thing due to assembly binding issues.

File: src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantWebHostBuilderExtensions.cs (hunk 4 above)
Comment (alsami) [reply src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantWebHostBuilderExtensions.cs]: What you elaborate why this would need to have the namespace of a different package?

File: src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs (hunk 6 above)
Comment (alsami) [reply src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs]: I added conditional references and imports. Can you please check if you think we can leave it that way? 

I'd suggest also to add more samples targeting older versions of ASP.NET Core to make sure it's really working (later, when the actual purpose is restored, ofc.)

File: src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj (hunk 5 above)
Comment (alsami) [reply src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj]: Readded support for ASP.NET Core 2.x

File: src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs (hunk 6 above)
Comment (tillig) [reply src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs]: I think the conditional import thing is fine. We've done that in other packages. I agree we should have a sandbox for each version to make sure things are working; IIRC we wanted that in the Autofac.Extensions.DependencyInjection repo for the same reason.

**Comment by tillig** [comment]
@alsami You might consider doing several commits locally and pushing when you're closer to ready to review time, or at least in larger chunks. GitHub fires off notifications every time there's a push to a pull request so it's hard to determine if it's something I should be coming back to address.

**Comment by alsami** [comment]
> I think the overall pattern of providing a callback to the multitenant factory to configure the multitenant container is good, but the execution of the callback needs to happen right in the factory and put into the returned AutofacServiceProvider rather than put into the middleware. This may mean there needs to be an AutofacMultitenantServiceProvider that derives from the main AutofacServiceProvider but extends it to add a reference to the root app container AND the multitenant container.
> 
> Since the "lifetime scope" passed to the provider will be the multitenant container, the additional property (and extension method) will probably need to be to get the root container.
> 
> If there's a better way, I'm open to suggestions - I'm not saying "a new service provider class must happen!", I'm saying that's one way to fix it.

I tried that in a different branch [here](https://github.com/alsami/Autofac.AspNetCore.Multitenant/blob/4a54045dc3d6130c7491adc33a836baf6c75d3b9/src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProvider.cs#L44) and using it [here](https://github.com/alsami/Autofac.AspNetCore.Multitenant/blob/4a54045dc3d6130c7491adc33a836baf6c75d3b9/src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs#L93).

From my understanding I should be able to down-cast an instance of `IServiceProvider` to `AutofacMultitenantServiceProvider` to retrieve the instance of `MultitenantContainer`. This is apparently not the case and I retrieve an instance of the base-class, `AutofacServiceProvider`. 

I tried it [here ](https://github.com/alsami/Autofac.AspNetCore.Multitenant/blob/0e2a39b59175bb1f7faba72a6c14ae45a68d6123/src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs#L55)and I get an `InvalidCastException`.

Maybe I got you wrong since I still haven't grasp all of the "magic" happening here. Would you mind checking it out? Bear in mind that the test-project is currently not working.

I think this is because the [HostBuilder](https://github.com/aspnet/Extensions/blob/f8d29bc1344364c7444bec17049df653f82760c7/src/Hosting/Hosting/src/HostBuilder.cs#L233) is creating the main instance of  the underlying `IServiceProvider` and keeps it for itself. Everything else is a child-scope.

I did that in a seperate branch because of the many irrelevant changes that have happened in the branch assigned for this PR. I was going to override the changes here with the ones I made on that new branch.

Thanks in advance and many thanks for the feedback and support so far @tillig.

**Comment by tillig** [comment]
The `IServiceProvider` being used in the middleware _should be_ the one built out of the app startup.

Browsing through I don't see anything in particular that would cause that exception (though I see there was a little [experimentation to remove the extension method to add the request services, which I'd avoid because it removes some control from the developer](https://github.com/alsami/Autofac.AspNetCore.Multitenant/blob/0e2a39b59175bb1f7faba72a6c14ae45a68d6123/src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs#L65)).

Did you try debugging that to see what the service provider was at the time? And make sure that the multitenant provider was created? If I have time I will try to check out the branch, but don't just sit and wait for that because I'm a little strapped (hence the original call for owners). What I'll be doing is throwing a bunch of breakpoints in and stepping through code, possibly stepping down into the hosting bits and trying to figure out the order of operations.

- When does the middleware get added?
- Is that before the multitenant service provider gets created?
- What type of service provider shows up in the middleware?
- If you resolve an IServiceProvider from the container, what do you get? Is it the same as the type of IServiceProvider that goes in the middleware? (I'm guessing this may have something to do with it.)

That's the stuff I'd be looking at.

**Comment by alsami** [comment]
I think I've found the course and it's us :-). 

The `Populate(IServiceCollection)` method registers the following
image

This also means that the `AutofacChildScopeProviderFactory` implemented in `Autofac.Extensions.DependencyInjection` is not working at all.

The only time we have the actual `IServiceProviderFactory` implementation used is in the `HostBuilder`.

image

Brings us a step closer. Mind checking my observations are correct, @tillig?

I will try to duplicate the logic for populating for now to see if we can work with that.

**Comment by tillig** [comment]
I don't have time at the moment to set up a repro and debug into this, but the description here sounds right.

In the AutofacMultitenantServiceProviderFactory you could always register the AutofacMultitenantServiceProvider in the builder _after_ the `Populate` call.

```c#
AutofacMultitenantServiceProvider provider = null;
builder.Populate(services);
builder.Register(ctx => mtc).As<IServiceProvider>().SingleInstance();
var mtc = new MultitenantContainer(strategy, builder.Build());
provider = new AutofacMultitenantServiceProvider(mtc);
```
That sort of thing. The idea being the closure adds a lazy reference to the correct service provider, it overrides the registration done in `Populate`, and you get the right thing out when you resolve.

**Comment by alsami** [comment]
That seems to be working.

**Comment by alsami** [comment]
@tillig good to go to review again.
Tests are passing for both frameworks.

image

Please let me know if there is anything that still needs to be done.

File: src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj [hunk 15]
```diff
@@ -41,10 +48,10 @@
     <PackageReference Include="Microsoft.Net.RoslynDiagnostics" Version="2.6.1">
       <PrivateAssets>All</PrivateAssets>
     </PackageReference>
-    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta-62925-02" Condition="Exists('$(MSBuildThisFileDirectory)../../.git')">
+    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-19367-01">
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj:51]: It appears the condition here got nuked. This will break folks who download the code in a zip file and try to build.

File: src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs [hunk 16]
```diff
@@ -1,11 +1,13 @@
 ﻿using System;
 using System.Linq;
 using System.Linq.Expressions;
-using System.Reflection;
-using Microsoft.AspNetCore.Hosting.Internal;
+using Autofac.Integration.AspNetCore.Multitenant.Properties;
 using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Http.Features;
 using Microsoft.Extensions.DependencyInjection;
+#if NETSTANDARD2_0
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs:8]: It looks like the changes made here could all probably be reverted - it doesn't seem there was actually anything that happened.

File: src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProvider.cs [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    /// Autofac implementation of the ASP.NET Core <see cref="IServiceProvider"/> for a <see cref="MultitenantContainer" />.
+    /// </summary>
+    /// <seealso cref="IServiceProvider" />
+    public sealed class AutofacMultitenantServiceProvider : AutofacServiceProvider
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutofacMultitenantServiceProvider"/> class.
+        /// </summary>
+        /// <param name="lifetimeScope">The <see cref="ILifetimeScope"/> in form of a <see cref="MultitenantContainer"/>.</param>
+        public AutofacMultitenantServiceProvider(ILifetimeScope lifetimeScope)
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProvider.cs:42]: This provider should take a MultitenantContainer as a constructor parameter because it's very specific to supporting the multitenant scenario. That container can be passed to the base constructor, and the base `AutofacServiceProvider.LifetimeScope` will then point to the multitenant container.

It might also be interesting to add an extension method like `GetAutofacMultitenantRoot()` where it calls `GetAutofacRoot()` and casts it automatically to the multitenant container. That could be used later in the middleware and would be useful for others to not repeat that block of code.

File: src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs [hunk 18]
```diff
@@ -52,8 +52,7 @@ public async Task Invoke(HttpContext context)
                 this._contextAccessor.HttpContext = context;
             }
 
-            var container = this._multitenantContainerAccessor();
-            if (container == null)
+            if (!(this._serviceProvider.GetAutofacRoot() is MultitenantContainer container))
```
Comment (tillig) [inline src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs:55]: If you have `GetAutofacMultitenantRoot()` as noted above, this would just be a null check.

**Review by tillig — CHANGES_REQUESTED** [review]
This is looking really, really super. 🎉 A couple of minor changes and I think we can call this a wrap. Great work, thanks for sticking with it!

File: src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs (hunk 16 above)
Comment (alsami) [reply src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs:8]: Did you try it? It gives me a compile error without it.
image

Comment (tillig) [reply src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs:8]: No, I didn't try it. If it needs to be there, it's fine to leave it. They must have moved some types around with that whole "pubternal" thing they were doing a while ago.

File: src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs (hunk 18 above)
Comment (tillig) [reply src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs:55]: I just noticed that GetAutofacRoot actually throws InvalidOperationException if it's not an Autofac service provider, so if GetAutofacMultitenantRoot also throws, you wouldn't even have to null check.

**Comment by alsami** [comment]
> This is looking really, really super.  A couple of minor changes and I think we can call this a wrap. Great work, thanks for sticking with it!

Appreciate the feedback and support mate.

Please go for the (hopefully last) review!

**Review by tillig — APPROVED** [review]
Looks great. We can add a `GetAutofacMultitenantRoot` extension later. Thanks again!

**Comment by alsami** [comment]
> Looks great. We can add a `GetAutofacMultitenantRoot` extension later. Thanks again!

Damn, I actually added it, sadly on the wrong branch -.-

https://github.com/alsami/Autofac.AspNetCore.Multitenant/blob/feat/aspnetcore3_support_rework/src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderExtensions.cs

MR incoming.
